remote.go

```diff
@@ -0,0 +1,265 @@
+			rd.Log.Debugf("failed to connect to %s: %v", addr, err)
+			lastErr = err
+			if !isTemporaryErr(err) {
+				break
+			}
+			continue
+		}
+		usedServer = addr
+	}
+	if cl == nil {
+		return toPartialError(isTemporaryErr(lastErr), rcpts, lastErr)
+	}
+
+	rd.Log.Debugln("connected to", usedServer)
+
+	if err := cl.Mail(ctx.From); err != nil {
+		rd.Log.Printf("MAIL FROM failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+
+	perr := PartialError{Errs: make(map[string]error)}
+	for _, rcpt := range rcpts {
+		if err := cl.Rcpt(rcpt); err != nil {
+			rd.Log.Printf("RCPT TO failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+			if isTemporaryErr(err) {
+				perr.TemporaryFailed = append(perr.TemporaryFailed, rcpt)
+			} else {
+				perr.Failed = append(perr.Failed, rcpt)
+			}
+			perr.Errs[rcpt] = err
+		}
+	}
+	bodyWriter, err := cl.Data()
+	if err != nil {
+		rd.Log.Printf("DATA failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+	defer bodyWriter.Close()
+	if _, err := io.Copy(bodyWriter, body); err != nil {
+		rd.Log.Printf("body write failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+
+	return nil
+}
+
+func connectToServer(ourHostname, address string, requireTLS bool) (*smtp.Client, error) {
+	cl, err := smtp.Dial(address + ":25")
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cl.Hello(ourHostname); err != nil {
+		return nil, err
+	}
+
+	if tlsOk, _ := cl.Extension("STARTTLS"); tlsOk {
+		if err := cl.StartTLS(&tls.Config{
+			ServerName: address,
+		}); err != nil {
+			return nil, err
+		}
+	} else if requireTLS {
+		return nil, ErrTLSRequired
+	}
+
+	return cl, nil
+}
+
+func groupByDomain(rcpts []string) (map[string][]string, error) {
+	res := make(map[string][]string, len(rcpts))
+	for _, rcpt := range rcpts {
+		parts := strings.Split(rcpt, "@")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("malformed address: %s", rcpt)
+		}
+
+		res[parts[1]] = append(res[parts[1]], rcpt)
+	}
+	return res, nil
+}
+
+func lookupTargetServers(domain string) ([]string, error) {
+	records, err := net.LookupMX(domain)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Slice(records, func(i, j int) bool {
+		return records[i].Pref < records[j].Pref
+	})
+
+	hosts := make([]string, 0, len(records))
+	for _, record := range records {
+		hosts = append(hosts, record.Host)
+	}
+	return hosts, nil
+}
+
+func isTemporaryErr(err error) bool {
+	if protoErr, ok := err.(*textproto.Error); ok {
+		return (protoErr.Code / 100) == 4
+	}
+	if smtpErr, ok := err.(*smtp.SMTPError); ok {
+		return (smtpErr.Code / 100) == 4
+	}
+	if dnsErr, ok := err.(*net.DNSError); ok {
+		return dnsErr.Temporary()
+	}
+
+	if strings.HasPrefix(err.Error(), "x509") {
```

Comment:
Ugh

Reply:
Errors with `x509` prefix indicate certificate verification failure. I'm not sure whether we should consider them to be temporary. Wdyt? Anyway, somehow I missed that there is a concrete type for it (`x509.CertificateInvalidError`).

Reply:
1. My "ugh" was about checking the error message. This should really be last-resort if we really need it.
2. Because of certificates expiring and/or misconfiguration, it's probably a good idea to retry anyway.

Reply:
Done. Removed this check.